File: frontend/src/api/courses.ts
import { api } from "./client";
import type { Course, CourseDetail, Enrollment } from "../types";

export async function getCourses(params?: {
  search?: string;
  level?: string;
  skip?: number;
  limit?: number;
}): Promise<Course[]> {
  const cleanParams = Object.fromEntries(
    Object.entries(params || {}).filter(([_, v]) => v !== undefined && v !== "")
  );
  const { data } = await api.get<Course[]>("/courses", { params: cleanParams });
  return data;
}

export async function getCourse(courseId: string): Promise<CourseDetail> {
  const { data } = await api.get<CourseDetail>(`/courses/${courseId}`);
  return data;
}

export async function enrollInCourse(courseId: string): Promise<Enrollment> {
  const { data } = await api.post<Enrollment>(`/courses/${courseId}/enroll`);
  return data;
}

export async function getMyEnrollments(): Promise<Enrollment[]> {
  const { data } = await api.get<Enrollment[]>("/courses/me/enrollments");
  return data;
}

File: backend/app/api/routes/courses.py
from uuid import UUID

from fastapi import APIRouter, HTTPException, Query, status

from app.api.deps import CurrentUser, DbSession
from app.schemas.course import CourseCreate, CourseDetail, CourseRead, EnrollmentRead
from app.services.courses import (
    create_course,
    enroll_user,
    get_course,
    get_course_with_lessons,
    list_courses,
    list_user_enrollments,
)

router = APIRouter()


@router.get("", response_model=list[CourseRead])
def read_courses(
    db: DbSession,
    search: str | None = Query(None, min_length=1, description="Search courses by title"),
    level: str | None = Query(None, description="Filter by level"),
    skip: int = Query(0, ge=0, description="Pagination offset"),
    limit: int = Query(50, ge=1, le=100, description="Pagination limit"),
) -> list[CourseRead]:
    courses, _ = list_courses(db, search=search, level=level, skip=skip, limit=limit)
    return courses


@router.get("/{course_id}", response_model=CourseDetail)
def read_course(course_id: UUID, db: DbSession) -> CourseDetail:
    course = get_course_with_lessons(db, course_id)
    if not course:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Course not found")
    return course


@router.post("", response_model=CourseRead, status_code=status.HTTP_201_CREATED)
def add_course(payload: CourseCreate, db: DbSession, current_user: CurrentUser) -> CourseRead:
    if not current_user.is_superuser:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Admin access required")
    return create_course(db, payload)


@router.post("/{course_id}/enroll", response_model=EnrollmentRead)
def enroll(course_id: UUID, db: DbSession, current_user: CurrentUser) -> EnrollmentRead:
    course = get_course(db, course_id)
    if not course:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Course not found")
    enrollment = enroll_user(db, current_user, course)
    enrollment.course = course
    return enrollment


@router.get("/me/enrollments", response_model=list[EnrollmentRead])
def my_enrollments(db: DbSession, current_user: CurrentUser) -> list[EnrollmentRead]:
    return list_user_enrollments(db, current_user)
